torch/nested/_internal/ops.py

```diff
@@ -884,16 +884,91 @@ def split_with_sizes_default(func, *args, **kwargs):
     ]
 
 
+# TODO: Implement slice() instead and narrow() in terms of slice()
 @register_jagged_func(
-    torch.ops.aten.narrow.default, "self: jt, dim: any, start: any, length: any"
+    torch.ops.aten.narrow.default, "self: jt_all, dim: any, start: any, length: any"
 )
 def narrow(func, *args, **kwargs):
     _, new_kwargs = normalize_function(  # type: ignore[misc]
         func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
     )
     inp = new_kwargs.pop("input")
 
-    dim = _wrap_jagged_dim(inp.dim(), new_kwargs["dim"], inp._ragged_idx, "narrow")
+    dim, operating_on_batch = _wrap_jagged_dim(
+        inp.dim(), new_kwargs["dim"], inp._ragged_idx, "narrow", allow_batch_dim=True
+    )
+    if operating_on_batch:
+        # batch dim narrowing requires custom logic involving offsets
+        out_kwargs = extract_kwargs(inp)
+        start_val, length_val = new_kwargs["start"], new_kwargs["length"]
+        end_val = start_val + length_val
+        batch = inp._offsets.shape[0] - 1
+        if end_val > batch:
+            raise RuntimeError(
+                f"narrow(): start ({start_val}) + length ({length_val}) "
+                f"exceeds dimension size ({batch})"
+            )
+
+        # clamp start, end values
+        if start_val < 0:
+            start_val += inp._values.size(dim)
+        if end_val < 0:
+            end_val += inp._values.size(dim)
+        start_val = max(min(start_val, inp._values.size(dim)), 0)
+        end_val = max(min(end_val, inp._values.size(dim)), 0)
+        length_val = max(min(length_val, end_val - start_val), 0)
```

Comment:
Unfortunate that we have to duplicate narrow input checking/manipulation logic here, but I guess maybe hard to avoid due to as_strided use.

Reply:
Yeah agreed, I don't like this at all :( but indeed we aren't redispatching to narrow(), so we can't utilize the checks there